### PR TITLE
Implement IMPEC API for customers & restaurants

### LINKED ISSUE #3680: Implement IMPEC - for customers

### PR DESCRIPTION
**Pull request by alexsegura**
Fixes #3680

It is now possible: 
- To connect a restaurant with its IMPEC account
- To setup IMPEC "formats" for each products
- To allow a customer to connect his/her IMPEC account
- For a restaurant to "amend" the quantity of "formats" used in the order when packing

#### Connecting for a restaurant

loopeat_connect_resto

#### Setting up formats for a product

loopeat_setup_formats

#### Checkout

loopeat_add_to_cart

#### Amending formats used by restaurant

loopeat_resto_order

**TODO**

- [x] Allow configuring service fee per initiative
- [x] Add popups to explain how it works

### CODE REVIEW

**Comment by biologeek** [comment]
Hi everyone, good to see that ! 

Looking at the videos, it looks easy to implement for our restaurants. Great job !
I didn't see any related change in the app. Is it also planned ? Mainly for the ability for restaurants to amend the formats at order validation.

File: src/Api/DataTransformer/UpdateLoopeatFormatsDataTransformer.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            // TODO Manage this via API / iriConverter
+            preg_match('#\/api\/orders\/(?<orderId>[0-9]+)\/items\/(?<itemId>[0-9]+)#',
+                $item->orderItem['@id'], $matches);
+
+            $itemId = $matches['itemId'];
+
+            $orderItem = $this->entityManager->getRepository(OrderItem::class)->find($itemId);
+
+            if (null === $orderItem) {
+                // TODO Throw 400
```
Comment (lucasferraro) [inline src/Api/DataTransformer/UpdateLoopeatFormatsDataTransformer.php:39]: This commnet about code 400 should be implemented now?

File: src/LoopEat/Client.php [hunk 2]
```diff
@@ -148,129 +166,305 @@ public function getOAuthAuthorizeUrl(array $params = [])
...
-    public function return(Customer $customer, $quantity = 1): bool
+    public function getRestaurantOAuthAuthorizeUrl()
     {
-        $this->logger->info(sprintf('Returning %d Loopeats from "%s"',
-            $quantity, $customer->getEmailCanonical()));
+        // TODO
+        // Retrieve it from /api/v1/partners/initiatives + cache results
 
-        if ($quantity < 1) {
+        return 'https://resto.green.preprod.collectif-impec.org/oauth/authorize';
```
Comment (lucasferraro) [inline src/LoopEat/Client.php:219]: This URL should be configurable?

File: templates/_partials/restaurant/menu_item_inner.html.twig [hunk 3]
```diff
@@ -10,8 +10,13 @@
...
       {% include '_partials/restaurant/menu_item_content.html.twig' %}
     </span>
-    <span class="menu-item-price">
-      {{ variant.price|price_format }}
+    <span class="d-flex align-items-center">
+      <span class="menu-item-price">
+        {{ variant.price|price_format }}
+      </span>
+      {% if product.isReusablePackagingEnabled() %}
+      <span class="label label-info ml-2">Zéro déchet</span>
```
Comment (lucasferraro) [inline templates/_partials/restaurant/menu_item_inner.html.twig:18]: translation

**Review by lucasferraro — CHANGES_REQUESTED** [review]
Added some minor suggestions and comments.

**Comment by alexsegura** [comment]
@biologeek for the app, it's here --> coopcycle/coopcycle-app#1521

File: src/Entity/Sylius/Order.php [hunk 4]
```diff
@@ -349,6 +356,43 @@
...
+ *       "input"=LoopeatFormatsOutput::class,
+ *       "validate"=false,
+ *       "normalization_context"={"groups"={"cart"}},
+ *       "denormalization_context"={"groups"={"update_loopeat_formats"}},
+ *       "openapi_context"={
+ *         "summary"="Update Loopeat formats for an order"
+ *       }
+ *     },
+ *     "update_loopeat_returns"={
+ *       "method"="POST",
```
Comment (lucasferraro) [inline src/Entity/Sylius/Order.php:384]: Should it be a "PUT" method like `update_loopeat_formats` ?

**Review by lucasferraro — CHANGES_REQUESTED** [review]

File: src/Entity/Sylius/Order.php (hunk 4 above)
Comment (alexsegura) [reply src/Entity/Sylius/Order.php:384]: Oops, the changes on the app have been merged & deployed, and it using POST

https://github.com/coopcycle/coopcycle-app/blob/6675e8a18177ca64bfe5fe79216087cf2358ad88/src/redux/Checkout/actions.js#L1360

**Pull request merged**